drivers/wifi/esp8266/esp8266.c

```diff
@@ -0,0 +1,844 @@
+
+	dev = CONTAINER_OF(data, struct esp8266_data, cmd_handler_data);
+	sock = esp8266_socket_from_link_id(dev, link_id);
+	if (sock == NULL) {
+		LOG_ERR("No socket for link %d", link_id);
+		return 0;
+	}
+
+	if (!esp8266_socket_connected(sock)) {
+		LOG_WRN("Link %d already closed", link_id);
+		return 0;
+	}
+
+	sock->flags &= ~(ESP_SOCK_CONNECTED);
+	k_work_submit_to_queue(&dev->workq, &sock->recv_work);
+
+	return 0;
+}
+
+/*
+ * Passive TCP: "+IPD,<id>,<len>\r\n"
+ * Other:       "+IPD,<id>,<len>:<data>"
+ */
+#define MIN_IPD_LEN (sizeof("+IPD,I,LE") - 1)
+#define MAX_IPD_LEN (sizeof("+IPD,I,LLLLE") - 1)
+MODEM_CMD_DIRECT_DEFINE(on_cmd_ipd)
+{
+	char *endptr, end, ipd_buf[MAX_IPD_LEN + 1];
+	int data_offset, data_len, read_len, ret;
+	size_t match_len, frags_len;
+	struct esp8266_socket *sock;
+	struct esp8266_data *dev;
+	struct net_pkt *pkt;
+	u8_t link_id;
+
+	dev = CONTAINER_OF(data, struct esp8266_data, cmd_handler_data);
+
+	esp8266_lock(dev);
+
+	frags_len = net_buf_frags_len(data->rx_buf);
+
+	/* Wait until minimum cmd length is available */
+	if (frags_len < MIN_IPD_LEN) {
+		ret = -EAGAIN;
+		goto out;
+	}
+
+	match_len = net_buf_linearize(ipd_buf, MAX_IPD_LEN,
+				      data->rx_buf, 0, MAX_IPD_LEN);
+
+	ipd_buf[match_len] = 0;
+	if (ipd_buf[len] != ',' || ipd_buf[len + 2] != ',') {
+		LOG_ERR("Invalid IPD: %s", log_strdup(ipd_buf));
+		ret = len;
+		goto out;
+	}
+
+	link_id = ipd_buf[len + 1] - '0';
+	sock = esp8266_socket_from_link_id(dev, link_id);
+	if (sock == NULL) {
+		LOG_ERR("No socket for link %d", link_id);
+		ret = len;
+		goto out;
+	}
+
+	/* When using passive TCP, the +IPD command ends with \r\n */
+	if (IS_ENABLED(CONFIG_WIFI_ESP8266_PASSIVE_TCP) &&
+	    sock->ip_proto == IPPROTO_TCP) {
+		end = '\r';
+	} else {
+		end = ':';
+	}
+
+	data_len = strtol(&ipd_buf[len + 3], &endptr, 10);
+	if (endptr == &ipd_buf[len + 3] ||
+	    (*endptr == 0 && match_len >= MAX_IPD_LEN)) {
+		/* Invalid */
+		LOG_ERR("Invalid IPD len: %s", log_strdup(ipd_buf));
+		ret = len;
+		goto out;
+	} else if (*endptr == 0) {
+		ret = -EAGAIN;
+		goto out;
+	} else if (*endptr != end) {
+		LOG_ERR("Invalid cmd end 0x%02x, expected 0x%02x", *endptr,
+			end);
+		ret = len;
+		goto out;
+	}
+
+	*endptr = 0;
+	data_offset = strlen(ipd_buf) + 1;
+
+	/*
+	 * When using passive TCP, the data itself is not included in the +IPD
+	 * command but must be polled with AT+CIPRECVDATA.
+	 */
+	if (IS_ENABLED(CONFIG_WIFI_ESP8266_PASSIVE_TCP) &&
+	    sock->ip_proto == IPPROTO_TCP) {
+		sock->bytes_avail = data_len;
+		k_work_submit_to_queue(&dev->workq, &sock->recvdata_work);
+		ret = data_offset;
+		goto out;
+	}
+
+	/* Do we have the whole message? */
+	if (data_offset + data_len > frags_len) {
+		ret = -EAGAIN;
+		goto out;
+	}
+
+	ret = data_offset + data_len; /* Skip */
+
+	pkt = net_pkt_rx_alloc_with_buffer(dev->net_iface, data_len, AF_UNSPEC,
+					   0, K_NO_WAIT);
+	if (!pkt) {
+		LOG_ERR("Could not allocate net_pkt");
+		goto out;
+	}
```

Comment:
Oddly found that I ran out of rx buffers very quickly. It may be because I was pushing a lot of small messages which it struggled with.

Reply:
Yea I've had that problem as well. I added a timeout for 100 ms which should give time for a thread to read socket and free up some `net_buf`'s. Also for downloading much data I increased `CONFIG_NET_BUF_RX_COUNT`.